## SquamataAssemballyAMT - Jupyter notebook for batch releasing Audio Magnetotellic (AMT) data to ScienceBase

This module performs the following operations:
- Create list of data directories.
- Identify files accompanying data release.
- Create file listing for metadata XML markup.
- Identify and load MT EDI file.
- Collect and harvest release parameters common to ALL metadata childs.
- Clean up and reformat harvested values to be XML metadata complient.
- Create User Editable Keywords Listing
- Create entity and attribute XML markup.
- Poplulate metadata template
- Validate metadata; create error log; create HTML and FGDC Text versions of the metadata.
- Create all child metadata files from first example created in previous steps. (In development)
- Perhaps upload files to ScienceBase (In development)
- Change ScienceBase parameters such as citation information, add orcid ids, add USGS CMS tags, etc. (In development)

### To execute a function/command select a cell and Hold-Shift + Press-Enter

**The 'r' signifies a string literal. Use for paths.**

Metadata wizard:  Advanced, Open In a jupyter Notebook?
Metadata Wizard 2.o from ScienceBase

In [1]:
# Phil Brown (pbrown@usgs.gov) 2019 Beta
# Working Python 3 Notebook used to facilitate the release of Audio Magnetotelluric (AMT) Data to ScienceBase.

In [1]:
# Test Cell
print ("Jupyter is working.") #To run this cell, hold down Shift and press Enter.

Jupyter is working.


In [1]:
# Load required Libraries
import sys
import os
import zipfile
import csv
#import pysb
import requests
import shutil
from shutil import copyfile
import zipfile
import datetime
import glob
from lxml import etree
import json
import pickle
import shutil
import fileinput
import json
import pandas as pd
import numpy as np
from IPython.core.display import display
from IPython.core.display import HTML
from lxml import etree
##from pymdwizard.core.xml_utils import XMLRecord
##from pymdwizard.core.xml_utils import XMLNode
import re
from ipywidgets import *
from IPython.display import display
from IPython.html.widgets import widgets
import datetime
import dateutil.parser

C:\Users\pbrown\AppData\Local\Continuum\anacondaNew\lib\site-packages\IPython\html.py:14: ShimWarning: The `IPython.html` package has been deprecated since IPython 4.0. You should import from `notebook` instead. `IPython.html.widgets` has moved to `ipywidgets`.
  "`IPython.html.widgets` has moved to `ipywidgets`.", ShimWarning)


# 1) Step One - Set Directory Paths
## Please set directory paths below
### Directory paths include
- Data Path: This is the path to the data, data structure should have a directory for each station
- Template Path: The path to the XML metadata template being used for the data.  This template should already include all information common to all child metadata files e.g. originators, larger work citation, etc.

In [2]:
#Set Data Paths - perhaps we'll get a user form to do this some day?
mtDataPath = r"C:\CurrentWork\DataReleases\Arkansas AMT data release" #The 'r' signifies a string literal. Use for paths.
mtMataDataTemplatePath = r"C:\CurrentWork\DataReleases\Arkansas AMT data release"
mtMataDataTemplateName = "MT-MetaData_TEMPLATE.xml"

In [3]:
#Check Paths for the fun of it
print ('The MT Data Path is: ' + '"' + mtDataPath + '"')
mtMataDataTemplatePath + "\\" + mtMataDataTemplateName

The MT Data Path is: "C:\CurrentWork\DataReleases\Arkansas AMT data release"


'C:\\CurrentWork\\DataReleases\\Arkansas AMT data release\\MT-MetaData_TEMPLATE.xml'

# 2) Step Two - Collect Common Parameters
## The first step is collect the information common to all child metadata sets
### Values Include:
- Data Release Title
    - Currently the title is collected from the EDI file; this title generally contains the station number.
- Data Release Originator(s)
- Larger Work Title
- Larger Work Originator(s)
- Larger Work URL
- Theme Keywords
- Location Keyword
- etc. etc

**As things stand, all information is either gathered from the EDI file unless it is common to all files; these common items are to included in the metadata template file upfront...**



## Now, let's explore our data. 
- What files do we have? 
- What files do we import values from?

In [4]:
#Review content in file explorer

In [5]:
#Produce directory listing of station (SB Object Children)
#Either set up the root directory with station subdirectories only or delete non-station directories from the list array
mtDataDirList = os.listdir(mtDataPath)
mtDataDirList

['AMT050',
 'AMT070',
 'AMT090',
 'AMT115',
 'AMT140',
 'AMT170',
 'MT-MetaData_TEMPLATE.xml',
 'MT-MetaData_TEMPLATE.xml.bak']

In [6]:
#Let's start with the first staion and check the result - we can then loop through the process 
#for the remaining stations in the list.
mtStationPath = mtDataPath + '\\' + mtDataDirList[0]
mtStationPath

'C:\\CurrentWork\\DataReleases\\Arkansas AMT data release\\AMT050'

In [7]:
#Look for EDI file to load
ediList = glob.glob(os.path.join(mtStationPath, '**/*MT*.edi'),  recursive=True)
ediPath = ediList[0]
ediList
print ('EDI File Path:\n' + ediPath)       

EDI File Path:
C:\CurrentWork\DataReleases\Arkansas AMT data release\AMT050\USA-Arkansas-Buffalo_River-2017-AMT050.edi


## Enter the information unique to this data set but common to all metadata files
### These include:
- Data Release Title
- Data Release Authors
- Theme Keywords
- Location Keywords

## After this step, information will be harvested from the MT EDI file. 
### These include:
- ProductId=USA-New_Mexico-Rio_Grande_Rift-San_Luis_Basin-2009-AMT01.edi
- ExternalUrl Url=https://doi.org/10.5066/F72F7MQ7
- Attachment Filename=https://pubs.usgs.gov/of/2011/1264/report/OF11-1264.pdf
- Survey Purpose Description: 
- Data Description:
- Citation Title=Audiomagnetotelluric data, Taos Plateau Volcanic Field, New Mexico
- Citation Authors=Chad E. Ailes, Brian D. Rodriguez
- Citation Year=2011
- YearCollected=2009
- Country=USA                                  
- Ellipsoid=Clarke 1866                                                          
- Location datum=NAD27 CONUS                                                     
- SITE LATITUDE=36.752985000                                                     
- SITE LONGITUDE=-105.560966167                                                  
- Elevation units="meters"=2608.00                                                                     
- Start=2009-07-21T19:52:03 UTC/GMT
- End=2009-07-21T20:34:20 UTC/GMT
- ProcessingTimeSeriesUsed:
         wp01A1.bp1                                                                     
         wp01A2.bp1                                                                     
         wp01A1.sd6                                                                     
         wp01A2.sd6                                                                     
         wp01A1.sd7                                                                     
         wp01A2.sd8                                                                     
         wp01A2_3.sd9 
- Entities and Attributes:
    - FREQUENCIES
    - IMPEDANCE ROTATION ANGLES
    - IMPEDANCES
    - TIPPER PARAMETERS
    - COMPUTED PARAMETERS


## Lets now import and index values from the EDI Files
- We need these values for the metadata template.  
- We also want to run stats on some of these values for the entity and attributes section

In [8]:
    #Load EDI File and Read It
    ediFile = open(ediPath, 'r')
    ediContent = ediFile.read()
    ediFile.close()
    print(ediContent)


>HEAD                                                                           
                                                                                
  DATAID="Buffalo Natl River"                                                   
  ACQBY=USGS                                                                    
  ACQDATE=2017-08-27
  STATE=Arkansas                                                                
  COUNTY=Newton                                                                 
  UNITS=M                                                                       
  STDVERS=1.0                                                                   
  PROGVERS=GEOTOOLS_2.3                                                         
  PROGDATE=09/16/94                                                             
                                                                                
>INFO   MAXLINES=1000                                                           
       

In [9]:
#Now assign values to the SB MetaDataWizard Template unknowns
list_ = ediContent.splitlines()
list_length = len (list_)

# there are probally easier ways to loop through the below but I like having it all hard coded upfront
# it's easire to track an change for me
# use the example below to extract additional parameters
# Not that all variables being collected are not necessarily used in populating the template.
# Values can be hardcoded into the metadata xml template and/or harvested from the edi file

for X in list_:
  if "ProductId" in X:
    productArray = X.split('=')
    productIdArray = productArray[1].split('.')
    productId = productIdArray[0]
    # We may want to reformat this are parse out this name further for use with a root name based on the Data Release Title?
    productId = productId.replace("-", " ")
    productId = productId.replace("_", " ")
    drTitle = productId
    print ('Child Title: ' + productId)
  if "ExternalUrl Url" in X:
    externalURLArray = X.split('=')
    externalURL = externalURLArray[1]
    print ('<onlink>: ' + externalURL)
  if "STATE" in X:
    stateArray = X.split('=')
    state = stateArray[1].replace('"', "") #remove quotes around state
    print ('State: ' + state)
  if "COUNTY" in X:
    countyArray = X.split('=')
    county = countyArray[1]
    print ('County: ' + county)
  if "Start" in X:
    startArray = X.split('=')
    start = startArray[1]
    print ('Start: ' + start)
  if "End" in X:
    endArray = X.split('=')
    end = endArray[1]
    print ('End: ' + end)
  if "Attachment Filename" in X and "http" in X:
    lgwrklinkArray = X.split('=')
    lgwrklink = lgwrklinkArray[1]
    print ('Attachment Filename Link: ' + lgwrklink)
  if "Citation Title" in X:
    citTitArray = X.split('=')
    citTit = citTitArray[1]
    print ('Citation Title: ' + citTit)
  if "Citation Authors" in X:
    citNamesArray = X.split('=')
    citAuthorsArray = citNamesArray[1].split(',')
    for author in citAuthorsArray:
     author = author.strip()
     print ('Author: '+ author)
  if "Citation Year" in X:
    citYearArray = X.split('=')
    citYear = citYearArray[1]
    print ('Citation Year: ' + citYear)
  if "YearCollected" in X:
    yearColArray = X.split('=')
    yearCol = yearColArray[1]
    print ('Year Collected: ' + yearCol)
  if "Ellipsoid" in X:
    ellipsoidArray = X.split('=')
    ellipsoid = ellipsoidArray[1]
    print ('Ellipsoid: ' + ellipsoid)
  if "Location datum" in X:
    locDatumArray = X.split('=')
    locDatum = locDatumArray[1]
    print ('Local datum: ' + locDatum)
  if "SITE LATITUDE" in X:
    sitLatArray = X.split('=')
    sitLat = sitLatArray[1] # !!! probally need to reformat this to have only 6 significant digits Also need to trim extra spaces !!!
    print ('Site latitude: ' + sitLat)
  if "SITE LONGITUDE" in X:
    sitLonArray = X.split('=')
    sitLon = sitLonArray[1] # !!! probally need to reformat this to have only 6 significant digits Also need to trim extra spaces !!!
    print ('Site longitude: ' + sitLon)
  if "Elevation units" in X:
    elevationStringArray = X.split('=')
    siteElevation = elevationStringArray[2] 
    print ('Site Elevation: ' + siteElevation)
    elevationUnits = elevationStringArray[1].replace('"', "")
    print ('Elevation Units: ' + elevationUnits)
    
# Code below returns values that occupy more than one line
    
for i in range(list_length):
 value = list_[i] 
 if value.replace(" ", "") == 'SurveyPurposeDescription:':
   startIndPurpose = i + 1
   #print ('startIndPurpose: ' + str(startIndPurpose))
 if value.replace(" ", "") == 'DataDescription:':
   endIndPurpose = i - 1
   #print ('endIndPurpose: ' + str(endIndPurpose))
purpose = list_[startIndPurpose]
for j in range(startIndPurpose + 1,endIndPurpose): 
    purpose = purpose + list_[j]
    purposeClean = re.sub(' +', ' ',purpose)
print ('\nAbstract:\n\t' + purposeClean)

for k in range(list_length):
 value = list_[k] 
 if value.replace(" ", "") == 'DataDescription:':
   startIndDescription = k + 1
   #print ('startIndDescription: ' + str(startIndDescription))
 if value.replace(" ", "") == 'FILECREATOR:':
   endIndDescription = k - 9
   #print ('endIndDescription: ' + str(endIndDescription))
description = list_[startIndDescription]
for l in range(startIndDescription + 1,endIndDescription): 
    description = description + list_[l]
    descriptionClean = re.sub(' +', ' ',description)
print ('\nPurpose:\n\t' + descriptionClean)
    

State: Arkansas                                                                
County: Newton                                                                 
Child Title: USA Arkansas Buffalo River 2017 AMT050
<onlink>: https://doi.org/10.5066/P9CIAXC5 
Citation Title: Audiomagnetotelluric data, Buffalo River watershed, Arkansas, 2017
Author: Brian D. Rodriguez and Mark R. Hudson
Citation Year: 2018                                                             
Year Collected: 2017
Ellipsoid: Clarke 1866                                                          
Local datum: NAD27 CONUS                                                     
Site latitude: 36.07789                                                         
Site longitude: -93.31091                                                       
Site Elevation: 681.91                                                
Elevation Units: meters
Start: 2017-08-27T19:01:59 UTC/GMT
End: 2017-08-27T23:45:56 UTC/GMT  
Start: 2017-08-27T19:01:59 

In [10]:
# Now let's format the start time and end time to be what the XML file wants for <begdate> and <enddate>

begdateArr = start.split(' ')
begdate_str = begdateArr[0]
begdate_obj = dateutil.parser.parse(begdate_str)
begdate = begdate_obj.strftime('%Y%m%d')
print('<begdate> ', begdate) 

enddateArr = end.split(' ')
enddate_str = enddateArr[0]
enddate_obj = dateutil.parser.parse(enddate_str)
enddate = enddate_obj.strftime('%Y%m%d')
print('<enddate> ', enddate) 


<begdate>  20170827
<enddate>  20170827


In [11]:
#Now we reformat the lat  and longitude to 6 sig figs as well as trim of any extra spaces
#Brian seems to be stripping out the sig figs now so there is no need.  
#Also may whant to round up instead of just stripping values?

sitLat = sitLat.strip()
##sitLat = sitLat[:-3] 
sitLon = sitLon.strip()
##sitLon = sitLon[:-3]

print ('Site latitude: ' + sitLat)
print ('Site longitude: ' + sitLon)

Site latitude: 36.07789
Site longitude: -93.31091


In [12]:
# Now Reformat county by trimming the extra spaces
county = county.strip()
print ('County: ' + county)

County: Newton


In [13]:
## Create editable keywords example.  
## Example text is created after running this cell
## This text is displayed by running "display(keywords) below
keywords = widgets.Textarea(
    value='\t\t<keywords>\n\t\t\t<theme>\n\t\t\t\t<themekt>ISO 19115 Topic Category</themekt>' \
    + '\n\t\t\t\t<themekey>biota</themekey>\n\t\t\t</theme>\n\t\t\t<theme>\n\t\t\t\t<themekt>None</themekt>' \
    + '\n\t\t\t\t<themekey>impedance</themekey>\n\t\t\t\t<themekey>tipper</themekey>' \
    + '\n\t\t\t\t<themekey>apparent resistivity</themekey>\n\t\t\t\t<themekey>impedance phase</themekey>' \
    + '\n\t\t\t\t<themekey>impedance strike</themekey>\n\t\t\t\t<themekey>MT</themekey>' \
    + '\n\t\t\t\t<themekey>audiomagnetotelluric</themekey>\n\t\t\t\t<themekey>magnetotelluric</themekey>' \
    + '\n\t\t\t\t<themekey>AMT</themekey>\n\t\t\t\t<themekey>sounding</themekey>' \
    + '\n\t\t\t\t<themekey>Geology, Geophysics, and Geochemistry Science Center</themekey>' \
    + '\n\t\t\t\t<themekey>GGGSC</themekey>\n\t\t\t\t<themekey>Mineral Resources Program</themekey>' \
    + '\n\t\t\t\t<themekey>MRP</themekey>\n\t\t\t</theme>\n\t\t\t<theme>\n\t\t\t\t<themekt>USGS Thesaurus</themekt>' \
    + '\n\t\t\t\t<themekey>Magnetic field (earth)</themekey>\n\t\t\t\t<themekey>Geophysics</themekey>' \
    + '\n\t\t\t\t<themekey>GPS measurement</themekey>\n\t\t\t\t<themekey>Electromagnetic surveying</themekey>' \
    + '\n\t\t\t\t<themekey>Magnetic surveying</themekey>\n\t\t\t</theme>\n\t\t\t<place>' \
    + '\n\t\t\t\t<placekt>USGS Geographic Names Information System (GNIS)</placekt>' \
    + '\n\t\t\t\t<placekey>New Mexico</placekey>\n\t\t\t\t<placekey>Rio Grande del Norte National Monument</placekey>' \
    + '\n\t\t\t\t<placekey>' + county + ' County</placekey>\n\t\t\t\t<placekey>Rio Grande</placekey>\n\t\t\t</place>\n\t\t</keywords>',
    placeholder='Type something',
    #description='String:',
    layout=Layout(width='100%', height='666px'),
    disabled=False
)
print ('Keywords list created.')

Keywords list created.


### Change the text in the textbox below to relflect what should be included as the key words for all child items

Note that changing the text below at any time creates a keywords section of the metadata seen EXACTLY as it is shown below

In [14]:
# Run this cell for key word text to edit.  
# Edit the text in place.  
# When complete move on to the next step

display(keywords)

Textarea(value='\t\t<keywords>\n\t\t\t<theme>\n\t\t\t\t<themekt>ISO 19115 Topic Category</themekt>\n\t\t\t\t<themekey>biota</themekey>\n\t\t\t</theme>\n\t\t\t<theme>\n\t\t\t\t<themekt>None</themekt>\n\t\t\t\t<themekey>impedance</themekey>\n\t\t\t\t<themekey>tipper</themekey>\n\t\t\t\t<themekey>apparent resistivity</themekey>\n\t\t\t\t<themekey>impedance phase</themekey>\n\t\t\t\t<themekey>impedance strike</themekey>\n\t\t\t\t<themekey>MT</themekey>\n\t\t\t\t<themekey>audiomagnetotelluric</themekey>\n\t\t\t\t<themekey>magnetotelluric</themekey>\n\t\t\t\t<themekey>AMT</themekey>\n\t\t\t\t<themekey>sounding</themekey>\n\t\t\t\t<themekey>Geology, Geophysics, and Geochemistry Science Center</themekey>\n\t\t\t\t<themekey>GGGSC</themekey>\n\t\t\t\t<themekey>Mineral Resources Program</themekey>\n\t\t\t\t<themekey>MRP</themekey>\n\t\t\t</theme>\n\t\t\t<theme>\n\t\t\t\t<themekt>USGS Thesaurus</themekt>\n\t\t\t\t<themekey>Magnetic field (earth)</themekey>\n\t\t\t\t<themekey>Geophysics</themekey>\n\t\t\t\t<themekey>GPS measurement</themekey>\n\t\t\t\t<themekey>Electromagnetic surveying</themekey>\n\t\t\t\t<themekey>Magnetic surveying</themekey>\n\t\t\t</theme>\n\t\t\t<place>\n\t\t\t\t<placekt>USGS Geographic Names Information System (GNIS)</placekt>\n\t\t\t\t<placekey>New Mexico</placekey>\n\t\t\t\t<placekey>Rio Grande del Norte National Monument</placekey>\n\t\t\t\t<placekey>Newton County</placekey>\n\t\t\t\t<placekey>Rio Grande</placekey>\n\t\t\t</place>\n\t\t</keywords>', layout=Layout(height='666px', width='100%'), placeholder='Type something')

Entity and Attribute Values for the EDI file.  List !****FREQUENCIES****!,!****IMPEDANCE ROTATION ANGLES****!,!****IMPEDANCES****!,!****COMPUTED PARAMETERS****!

Here we load the frequencies
>!****FREQUENCIES****!

In [15]:
# Import entity and attributes - !****FREQUENCIES****! plan to break some of these individual chunks into objects/functions

# Get Range of Frequency Values in EDI File
for k in range(list_length):

 value = list_[k] 
 if value.replace(" ", "") == '>!****FREQUENCIES****!':
   startIndFrequencies = k + 3
   print ('startIndFrequencies: ' + str(startIndFrequencies))
 
 if value.replace(" ", "") ==  '>!****IMPEDANCEROTATIONANGLES****!':
   endIndFrequencies = k - 1
   print ('endIndFrequencies: ' + str(endIndFrequencies))

frequencyData = []
fdata = []
fdataTemp = []
#Constuct a library of Headers and Values using Pandas, https://pandas.pydata.org/
frequencyDF = pd.DataFrame(fdata)
for j in range(startIndFrequencies,endIndFrequencies):
    fdataTemp = list_[j]
    fdataTemp = re.sub(' +', ' ',fdataTemp)
    fdataTemp = fdataTemp.split(" ")
    del fdataTemp[0]
    fdata = fdata + fdataTemp
    
print (fdata)  
fdata = np.array(fdata).astype(np.float) #convert String to floats
frequencyDF = pd.DataFrame(fdata,columns=['Frequencies'])
frequencyDF


startIndFrequencies: 277
endIndFrequencies: 283
['3.55000000E+03', '2.73000000E+03', '2.42000000E+03', '1.87000000E+03', '1.17000000E+03', '9.60000000E+02', '8.85000000E+02', '7.20000000E+02', '5.80000000E+02', '4.60000000E+02', '3.40000000E+02', '2.70000000E+02', '2.10000000E+02', '1.72399994E+02', '1.50000000E+02', '1.22099998E+02', '1.00000000E+02', '8.59400024E+01', '7.90000000E+01', '6.00600014E+01', '4.15000000E+01', '2.83199997E+01', '1.90400009E+01', '1.22100000E+01', '7.32399988E+00', '4.39400005E+00']


Frequencies
0   3550.000000
1   2730.000000
2   2420.000000
3   1870.000000
4   1170.000000
5    960.000000
6    885.000000
7    720.000000
8    580.000000
9    460.000000
10   340.000000
11   270.000000
12   210.000000
13   172.399994
14   150.000000
15   122.099998
16   100.000000
17    85.940002
18    79.000000
19    60.060001
20    41.500000
21    28.320000
22    19.040001
23    12.210000
24     7.324000
25     4.394000

In [16]:
# Now lets get the stats of the frequency data
#Make Array of Max Vallues
frequencyMax = frequencyDF[('Frequencies')].max()
print ('Max. Frequency: ' + str(frequencyMax))
frequencyMin = frequencyDF[('Frequencies')].min()
print ('Min. Frequency: ' + str(frequencyMin))

Max. Frequency: 3550.0
Min. Frequency: 4.39400005


Here we load the Impedance Rotation Angles
>!****IMPEDANCE ROTATION ANGLES****!

In [17]:
# Import entity and attributes - !****IMPEDANCE ROTATION ANGLES****! plan to break some of these individual chunks into objects/functions

# Get Range of Frequency Values in EDI File
for k in range(list_length):
 value = list_[k] 
 if value.replace(" ", "") == '>!****IMPEDANCEROTATIONANGLES****!':
   startIndROT = k + 3
   print ('startIndROT: ' + str(startIndROT))
 
 if value.replace(" ", "") ==  '>!****IMPEDANCES****!':
   endIndROT = k - 1
   print ('endIndROT: ' + str(endIndROT))

rdata = []
rdataTemp = []
#Constuct a library of Headers and Values using Pandas, https://pandas.pydata.org/
rotationDF = pd.DataFrame(rdata)
for j in range(startIndROT,endIndROT):
    rdataTemp = list_[j]
    rdataTemp = re.sub(' +', ' ',rdataTemp)
    rdataTemp = rdataTemp.split(" ")
    del rdataTemp[0]
    rdata = rdata + rdataTemp
    
print (rdata)  
rdata = np.array(rdata).astype(np.float) #convert String to floats
rotationDF = pd.DataFrame(rdata,columns=['ZROT'])
rotationDF

startIndROT: 287
endIndROT: 293
['0.00000000E+00', '0.00000000E+00', '0.00000000E+00', '0.00000000E+00', '0.00000000E+00', '0.00000000E+00', '0.00000000E+00', '0.00000000E+00', '0.00000000E+00', '0.00000000E+00', '0.00000000E+00', '0.00000000E+00', '0.00000000E+00', '0.00000000E+00', '0.00000000E+00', '0.00000000E+00', '0.00000000E+00', '0.00000000E+00', '0.00000000E+00', '0.00000000E+00', '0.00000000E+00', '0.00000000E+00', '0.00000000E+00', '0.00000000E+00', '0.00000000E+00', '0.00000000E+00']


ZROT
0    0.0
1    0.0
2    0.0
3    0.0
4    0.0
5    0.0
6    0.0
7    0.0
8    0.0
9    0.0
10   0.0
11   0.0
12   0.0
13   0.0
14   0.0
15   0.0
16   0.0
17   0.0
18   0.0
19   0.0
20   0.0
21   0.0
22   0.0
23   0.0
24   0.0
25   0.0

In [18]:
# Now lets get the stats of the rotation data
#Make Array of Max Vallues
rotationMax = rotationDF[('ZROT')].max()
print ('Max. ZROT: ' + str(frequencyMax))
rotationMin = rotationDF[('ZROT')].min()
print ('Min. ZROT: ' + str(frequencyMin))

Max. ZROT: 3550.0
Min. ZROT: 4.39400005


Here we load the impedances
>!****IMPEDANCES****!

In [19]:
# Import entity and attributes - !****IMPEDANCES****! plan to break some of these individual chunks into objects/functions

# Get Range of Impedance Values in EDI File
for k in range(list_length):
 value = list_[k] 
 if value.replace(" ", "") == '>!****IMPEDANCES****!':
   startIndImpedances = k + 1
   print ('startIndImpedances: ' + str(startIndImpedances))
 
 if value.replace(" ", "") ==  '>!****TIPPERPARAMETERS****!':
   endIndImpedances = k - 1
   print ('endIndImpedances: ' + str(endIndImpedances))

#Construct Array of Channel Headers   
count = 0
impedanceLabel = []
impedanceData = []
data = []
#Constuct a library of Headers and Values using Pandas, https://pandas.pydata.org/
impedanceDF = pd.DataFrame(data)
for l in range(startIndImpedances,endIndImpedances): 
    if list_[l][0] == '>':
     temp = list_[l].split(" ", 1)
     #print (temp)
     impedanceLabel.append((temp[0].split(">"))[1])
     dataTemp = list_[l+1]
     for j in range(l+2,l+8):
      dataTemp = dataTemp + list_[j]
      dataTemp = re.sub(' +', ' ',dataTemp)
     data = dataTemp.split(" ")
     del data[0] # need to check for empty strings and delete these from the array of the string can't be converted to a float
     del data[len(data)-1] # need to check for empty strings and delete
     #print (data)
     data = np.array(data).astype(np.float) #convert String to floats
     se = pd.Series(data)
     print ((temp[0].split(">"))[1])   
     impedanceDF[((temp[0].split(">"))[1])] = se.values
    
    count = count + 1

#impedanceDF = pd.DataFrame(data, columns=(impedanceLabel))
impedanceDF
#data
#se 

startIndImpedances: 295
endIndImpedances: 402
ZXXR
ZXXI
ZXX.VAR
ZXYR
ZXYI
ZXY.VAR
ZYXR
ZYXI
ZYX.VAR
ZYYR
ZYYI
ZYY.VAR


ZXXR        ZXXI      ZXX.VAR         ZXYR         ZXYI  \
0   367.010925  331.371613  7226.539550 -1093.219730  -657.220154   
1   306.581604  495.610596   844.390137  -341.106293   112.650757   
2   323.175903  325.950500   500.675842   115.803909   119.785774   
3   141.722778  442.028656   472.750183   393.872955  -545.837524   
4  -266.792267  686.778320  1314.924190  1229.474850 -1107.367430   
5  -319.259308 -205.590057  1086.698610  1152.989010   569.431763   
6   -61.971527  -35.475929   833.264771    61.100998   126.169968   
7   -42.160847 -246.904785   130.138062   707.872803   823.747742   
8    15.102742  -38.368511  1279.590940   171.901123   638.413269   
9    24.236101    0.444963   252.236435   245.412659   300.017944   
10   34.045483   28.716854    46.530338   262.411316   435.465149   
11  -39.249081  105.129677   347.813843   197.185394   134.824142   
12  -47.224739  111.250702   581.604187   146.170364   174.529388   
13  -16.372328   53.954666    62.758530   146.641159   108.860962   
14  -21.416239   33.042061    35.920574   146.738251    63.766819   
15  -21.185125   15.058688   145.250061   166.242569    45.572369   
16  -58.100662    5.489585   113.886604   155.369751    34.971756   
17    2.371945   11.134856    67.113411   111.825470    84.314758   
18   54.622177   -3.482544   273.899048   116.979309    54.943142   
19  -46.411930  -32.100685   114.546928   150.361710    44.213203   
20  -62.103477  -29.676872   271.616760    80.240044    54.353394   
21  -38.335106  -31.395382    45.463413   119.700554    38.115910   
22  -39.339787  -40.952961    20.383863    99.214386    26.583633   
23  -77.274795  -54.005764   121.705994    98.209518    84.220528   
24  -25.000242  -44.859775    36.347469   132.705444    14.589987   
25  -48.863194  -12.958874   287.609192    68.334915    29.106031   
26  -53.613361  -32.496223   258.740906    64.878510    20.970507   
27  -35.491306   -3.171625   130.494781    49.400948    17.109005   
28  -39.169182  -10.005348    46.368996    45.856918    13.728836   
29  -21.041618   12.445285   211.808685    53.507088    17.277393   

         ZXY.VAR         ZYXR        ZYXI       ZYX.VAR         ZYYR  \
0   18359.441400  -391.723572 -517.906982  38721.398400 -1675.782710   
1    8051.266110  -753.217468 -686.681396   2515.088130  -697.920837   
2    3154.015870  -882.499756 -499.462219   1389.263060   321.956177   
3    2136.897220  -978.120911 -329.152039    622.111145   548.287231   
4    8289.169920 -1298.830690 -305.449310   1112.016240  1123.639530   
5    3532.706540  -595.929382 -504.314697    569.315613   615.076538   
6    2252.473880  -182.721603 -147.348770   1041.355100    32.716114   
7     456.483582  -508.417694 -536.096191    278.142944   332.330872   
8    1370.859990   -98.669571  -56.908051    457.213959     4.779356   
9     312.407532  -317.878723 -425.340179    510.252533   148.002792   
10    126.007080  -285.025970 -485.368988     97.473587   142.499039   
11    137.210999  -140.565781 -175.290649    783.412415   232.732162   
12     69.032425  -196.781921 -129.315262   1557.326290   205.831741   
13     17.980564  -313.425446 -151.363281    360.592804   289.571594   
14      8.302307  -231.731522  -61.399719    228.813446   322.111267   
15      5.968705  -264.641388  -21.454197    523.724915   272.064941   
16      4.198465  -173.644058   49.037594    658.374695   285.082031   
17     15.852302  -199.378052  -84.385902    346.466644    79.025200   
18     27.412523   -55.732410   27.880737   2015.396360   143.300903   
19     24.432962  -134.561432  -15.031679    390.157379   150.329178   
20     66.050552  -122.637177   45.849056    955.710815    89.704559   
21     37.329903  -121.098801  -12.671127    110.778030    96.878036   
22     27.644882  -123.043846   11.834403     48.140713    88.423622   
23    276.661652  -133.414673   26.766451     87.716515   124.353142   
24    196.729630  -137.529022   71.543976    106.029793   173.791580  

In [20]:
# Now lets get the stats of the impedance data
#Make Array of Max Vallues
impedanceMax = []
for i in range (0,len(impedanceLabel)):
    impedanceMax.append(impedanceDF[(impedanceLabel[i])].max())
    
impedanceMin = []
for i in range (0,len(impedanceLabel)):
    impedanceMin.append(impedanceDF[(impedanceLabel[i])].min())

impedanceMin

[-319.259308,
 -246.904785,
 20.3838634,
 -1093.21973,
 -1107.36743,
 4.19846487,
 -1298.83069,
 -686.681396,
 48.1407127,
 -1675.78271,
 -1050.82922,
 21.5212269]

Here we load the tipper parameters
>!****TIPPER PARAMETERS****!

In [21]:
# Import entity and attributes - !****TIPPER PARAMETERS****! plan to break some of these individual chunks into objects/functions
# Probably will need two functions for this - one for a single list and one for the long lists with more than one column
# Get Range of Impedance Values in EDI File
for k in range(list_length):
 value = list_[k] 
 if value.replace(" ", "") == '>!****TIPPERPARAMETERS****!':
   startIndTipper = k + 1
   print ('startIndTipper: ' + str(startIndTipper))
 
 if value.replace(" ", "") ==  '>!****COMPUTEDPARAMETERS****!':
   endIndTipper = k - 1
   print ('endIndTipper: ' + str(endIndTipper))

#Construct Array of Channel Headers   
count = 0
tipperLabel = []
tipperData = []
tdata = []
#Constuct a library of Headers and Values using Pandas, https://pandas.pydata.org/
tipperDF = pd.DataFrame(tdata)
for l in range(startIndTipper,endIndTipper): 
    if list_[l][0] == '>':
     ttemp = list_[l].split(" ", 1)
     #print (ttemp)
     tipperLabel.append((ttemp[0].split(">"))[1])
     tdataTemp = list_[l+1]
     for j in range(l+2,l+8):
      tdataTemp = tdataTemp + list_[j]
      tdataTemp = re.sub(' +', ' ',tdataTemp)
      tdata = tdataTemp.split(" ")
     #print (tdata)
     del tdata[0]
     del tdata[len(tdata)-1] # need to check for empty strings and delete
     tdata = np.array(tdata).astype(np.float) #convert String to floats
     te = pd.Series(tdata)
     print ((ttemp[0].split(">"))[1])   
     tipperDF[((ttemp[0].split(">"))[1])] = te.values
    
    count = count + 1

#tipperDF = pd.DataFrame(tdata, columns=(tipperLabel))
tipperDF
#tdata
#te 

startIndTipper: 404
endIndTipper: 457
TXR.EXP
TXI.EXP
TXVAR.EXP
TYR.EXP
TYI.EXP
TYVAR.EXP


TXR.EXP   TXI.EXP  TXVAR.EXP   TYR.EXP   TYI.EXP  TYVAR.EXP
0   0.134743  0.023227   0.006163  0.815096  0.106846   0.015658
1  -0.374880  0.127913   0.006292  0.189078 -0.049881   0.059999
2   0.283068  0.127080   0.004819 -0.991411  0.018533   0.030358
3   0.194199 -0.046293   0.000132 -0.221568 -0.281418   0.000598
4   0.092244  0.132269   0.001762 -0.032708 -0.502063   0.011108
5   0.165504 -0.107432   0.000227  0.338185  0.004927   0.000737
6   0.390280 -0.108327   0.005069  0.976413 -0.251144   0.013702
7   0.170524 -0.100838   0.000263  0.162084  0.065559   0.000921
8   0.370160  0.116904   0.011584  0.300582 -0.004274   0.012411
9   0.254276  0.012799   0.001691  0.281947 -0.039079   0.002094
10  0.177984  0.100113   0.000652  0.082250  0.034320   0.001765
11  0.116498 -0.121757   0.007203  0.302602 -0.195460   0.002842
12  0.108679  0.063785   0.007914  0.208542 -0.143411   0.000939
13 -0.091268  0.097960   0.001335  0.284339 -0.284523   0.000383
14  0.262159 -0.184390   0.004406  0.408891 -0.503127   0.001018
15  0.085725 -0.024283   0.009138  0.316842 -0.496561   0.000375
16  0.313968 -0.213961   0.015426  0.398151 -0.574375   0.000569
17  0.049325 -0.011198   0.009340  0.086870 -0.023813   0.002206
18  0.967868 -0.745059   0.112085  0.307170 -0.318581   0.011218
19  0.344671 -0.693505   0.038562  0.354716 -0.487859   0.008225
20  0.694374 -0.696078   0.041616  0.125798 -0.143640   0.010120
21  0.392473 -0.613542   0.007552  0.298182 -0.102829   0.006201
22  0.552065 -0.741139   0.004748  0.364788 -0.152613   0.006439
23  0.607868 -0.783788   0.007499  0.560453 -0.300800   0.017048
24  0.859951 -0.869680   0.072374  0.552422 -0.637328   0.391724
25  1.349730 -0.857658   0.118802  0.029234 -0.155670   0.078286
26  1.622641 -0.976215   0.162638  0.029385  0.028866   0.114119
27  2.086255 -1.720377   0.259705  0.276317 -0.442084   0.257971
28  2.717031 -2.056379   0.148652  0.256761 -0.457995   0.509987
29  1.203977 -1.263167   1.498731 -0.030693 -0.151564   0.740546

In [22]:
# Now lets get the stats of the tipper data

# Make Array of Max Values
tipperMax = []
for i in range (0,len(tipperLabel)):
    tipperMax.append(tipperDF[(tipperLabel[i])].max())
print ('Tipper Max: ' + str(tipperMax))    

# Make Array of Min Values
tipperMin = []
for i in range (0,len(tipperLabel)):
    tipperMin.append(tipperDF[(tipperLabel[i])].min())
print ('Tipper Min: ' + str(tipperMin))

Tipper Max: [2.717031, 0.132268548, 1.49873149, 0.976413071, 0.106846295, 0.740545928]
Tipper Min: [-0.374880165, -2.05637932, 0.000132251458, -0.991410553, -0.637328267, 0.000375486881]


Here we load the computed parameters
>!****COMPUTED PARAMETERS****!

In [23]:
# Import entity and attributes - !****COMPUTED PARAMETERS****! plan to break some of these individual chunks into objects/functions
# Probably will need two functions for this - one for a single list and one for the long lists with more than one column
# Get Range of Impedance Values in EDI File
for k in range(list_length):
 value = list_[k] 
 if value.replace(" ", "") == '>!****COMPUTEDPARAMETERS****!':
   startIndPar = k + 1
   print ('startIndPar: ' + str(startIndPar))
 
 if value.replace(" ", "") ==  '>END':
   endIndPar = k - 1
   print ('endIndPar: ' + str(endIndPar))

#Construct Array of Channel Headers   
count = 0
parLabel = []
parData = []
pdata = []
#Constuct a library of Headers and Values using Pandas, https://pandas.pydata.org/
parDF = pd.DataFrame(pdata)
for l in range(startIndPar,endIndPar): 
    if list_[l][0] == '>':
     ptemp = list_[l].split(" ", 1)
     #print (ptemp)
     parLabel.append((ptemp[0].split(">"))[1])
     pdataTemp = list_[l+1]
     for j in range(l+2,l+8):
      pdataTemp = pdataTemp + list_[j]
      pdataTemp = re.sub(' +', ' ',pdataTemp)
      pdata = pdataTemp.split(" ")
     #print (pdata)
     del pdata[0]
     del pdata[len(pdata)-1] # need to check for empty strings and delete
     pdata = np.array(pdata).astype(np.float) #convert String to floats
     pe = pd.Series(pdata)
     print ((ptemp[0].split(">"))[1])   
     parDF[((ptemp[0].split(">"))[1])] = te.values
    
    count = count + 1

parDF
#pdata
#pe 

startIndPar: 459
endIndPar: 818
RHOROT
RHOXX
RHOXX.ERR
RHOXY
RHOXY.ERR
RHOYX
RHOYX.ERR
RHOYY
RHOYY.ERR
PHSXX
PHSXX.ERR
PHSXY
PHSXY.ERR
PHSYX
PHSYX.ERR
PHSYY
PHSYY.ERR
TIPMAG
TIPMAG.ERR
TIPPHS
TIPPHS.ERR
ZSTRIKE
ZSKEW
TSTRIKE
COH
COH
COH
COH
EPREDCOH
EPREDCOH
SIGAMP
SIGAMP
SIGAMP
SIGAMP
SIGAMP
SIGNOISE
SIGNOISE
SIGNOISE
SIGNOISE
SIGNOISE


RHOROT     RHOXX  RHOXX.ERR     RHOXY  RHOXY.ERR     RHOYX  RHOYX.ERR  \
0   0.015658  0.015658   0.015658  0.015658   0.015658  0.015658   0.015658   
1   0.059999  0.059999   0.059999  0.059999   0.059999  0.059999   0.059999   
2   0.030358  0.030358   0.030358  0.030358   0.030358  0.030358   0.030358   
3   0.000598  0.000598   0.000598  0.000598   0.000598  0.000598   0.000598   
4   0.011108  0.011108   0.011108  0.011108   0.011108  0.011108   0.011108   
5   0.000737  0.000737   0.000737  0.000737   0.000737  0.000737   0.000737   
6   0.013702  0.013702   0.013702  0.013702   0.013702  0.013702   0.013702   
7   0.000921  0.000921   0.000921  0.000921   0.000921  0.000921   0.000921   
8   0.012411  0.012411   0.012411  0.012411   0.012411  0.012411   0.012411   
9   0.002094  0.002094   0.002094  0.002094   0.002094  0.002094   0.002094   
10  0.001765  0.001765   0.001765  0.001765   0.001765  0.001765   0.001765   
11  0.002842  0.002842   0.002842  0.002842   0.002842  0.002842   0.002842   
12  0.000939  0.000939   0.000939  0.000939   0.000939  0.000939   0.000939   
13  0.000383  0.000383   0.000383  0.000383   0.000383  0.000383   0.000383   
14  0.001018  0.001018   0.001018  0.001018   0.001018  0.001018   0.001018   
15  0.000375  0.000375   0.000375  0.000375   0.000375  0.000375   0.000375   
16  0.000569  0.000569   0.000569  0.000569   0.000569  0.000569   0.000569   
17  0.002206  0.002206   0.002206  0.002206   0.002206  0.002206   0.002206   
18  0.011218  0.011218   0.011218  0.011218   0.011218  0.011218   0.011218   
19  0.008225  0.008225   0.008225  0.008225   0.008225  0.008225   0.008225   
20  0.010120  0.010120   0.010120  0.010120   0.010120  0.010120   0.010120   
21  0.006201  0.006201   0.006201  0.006201   0.006201  0.006201   0.006201   
22  0.006439  0.006439   0.006439  0.006439   0.006439  0.006439   0.006439   
23  0.017048  0.017048   0.017048  0.017048   0.017048  0.017048   0.017048   
24  0.391724  0.391724   0.391724  0.391724   0.391724  0.391724   0.391724   
25  0.078286  0.078286   0.078286  0.078286   0.078286  0.078286   0.078286   
26  0.114119  0.114119   0.114119  0.114119   0.114119  0.114119   0.114119   
27  0.257971  0.257971   0.257971  0.257971   0.257971  0.257971   0.257971   
28  0.509987  0.509987   0.509987  0.509987   0.509987  0.509987   0.509987   
29  0.740546  0.740546   0.740546  0.740546   0.740546  0.740546   0.740546   

       RHOYY  RHOYY.ERR     PHSXX    ...     TIPMAG.ERR    TIPPHS  TIPPHS.ERR  \
0   0.015658   0.015658  0.015658    ...       0.015658  0.015658    0.015658   
1   0.059999   0.059999  0.059999    ...       0.059999  0.059999    0.059999   
2   0.030358   0.030358  0.030358    ...       0.030358  0.030358    0.030358   
3   0.000598   0.000598  0.000598    ...       0.000598  0.000598    0.000598   
4   0.011108   0.011108  0.011108    ...       0.011108  0.011108    0.011108   
5   0.000737   0.000737  0.000737    ...       0.000737  0.000737    0.000737   
6   0.013702   0.013702  0.013702    ...       0.013702  0.013702    0.013702   
7   0.000921   0.000921  0.000921    ...       0.000921  0.000921    0.000921   
8   0.012411   0.012411  0.012411    ...       0.012411  0.012411    0.012411   
9   0.002094   0.002094  0.002094    ...       0.002094  0.002094    0.002094   
10  0.001765   0.001765  0.001765    ...       0.001765  0.001765    0.001765   
11  0.002842   0.002842  0.002842    ...       0.002842  0.002842    0.002842   
12  0.000939   0.000939  0.000939    ...       0.000939  0.000939    0.000939   
13  0.000383   0.000383  0.000383    ...       0.000383  0.000383    0.000383   
14  0.001018   0.001018  0.001018    ...       0.001018  0.001018    0.001018   
15  0.000375   0.000375  0.000375    ...       0.000375  0.000375    0.000375   
16  0.000569   0.000569  0.000569    ...       0.000569  0.000569    0.000569   
17  0.002206   0.002206  0.002206    ...       0.002206  0.002206    0.002206   
18  0.011218   0.

In [24]:
# Now lets get the stats of the computed parameters

# Make Array of Max Values
parMax = []
for i in range (0,len(parLabel)):
    parMax.append(parDF[(parLabel[i])].max())
print ('Computed Pararmeters Max: ' + str(parMax))    

# Make Array of Min Values
parMin = []
for i in range (0,len(parLabel)):
    parMin.append(parDF[(parLabel[i])].min())
print ('Computed Pararmeters Min: ' + str(parMin))

Computed Pararmeters Max: [0.740545928, 0.740545928, 0.740545928, 0.740545928, 0.740545928, 0.740545928, 0.740545928, 0.740545928, 0.740545928, 0.740545928, 0.740545928, 0.740545928, 0.740545928, 0.740545928, 0.740545928, 0.740545928, 0.740545928, 0.740545928, 0.740545928, 0.740545928, 0.740545928, 0.740545928, 0.740545928, 0.740545928, 0.740545928, 0.740545928, 0.740545928, 0.740545928, 0.740545928, 0.740545928, 0.740545928, 0.740545928, 0.740545928, 0.740545928, 0.740545928, 0.740545928, 0.740545928, 0.740545928, 0.740545928, 0.740545928]
Computed Pararmeters Min: [0.000375486881, 0.000375486881, 0.000375486881, 0.000375486881, 0.000375486881, 0.000375486881, 0.000375486881, 0.000375486881, 0.000375486881, 0.000375486881, 0.000375486881, 0.000375486881, 0.000375486881, 0.000375486881, 0.000375486881, 0.000375486881, 0.000375486881, 0.000375486881, 0.000375486881, 0.000375486881, 0.000375486881, 0.000375486881, 0.000375486881, 0.000375486881, 0.000375486881, 0.000375486881, 0.00037548

## Now lets get the range of values from the RSP values

In [25]:
#First Get the list of RSP files
rspList = glob.glob(os.path.join(mtStationPath, '*.RSP'),  recursive=True)
#rspList
fileListing = ''
rspFileListing = []
for i in range(len(rspList)):
    splitRspList = rspList[i].split('\\')
    fileListing = fileListing + '\t\t\t\t\t\t' + splitRspList[len(splitRspList) - 1] + '\n'
    rspFileListing.append(splitRspList[len(splitRspList) - 1])
print (fileListing)


						BF6-9621.RSP
						BF6-9624.RSP
						BF6-9625.RSP
						EF-9515X.RSP
						EF-9515Y.RSP



## Now the raw Binary File Listing - this can be T files or W files
We will need to figure out the best way of filtering on thise - may need to build array and then delete AVG, dmp and edi file.

These are listed in the edi file as well but they are not all there.  

    ProcessingTimeSeriesUsed:
         wp01A1.bp1                                                                     
         wp01A2.bp1                                                                     
         wp01A1.sd6                                                                     
         wp01A2.sd6                                                                     
         wp01A1.sd7                                                                     
         wp01A2.sd8                                                                     
         wp01A2_3.sd9 

- Which files need to be included in the data release?
- What is the best way to get this listing?

In [26]:
#Add the raw frequency files to the list except the AVG, dmp and edi file
binList = glob.glob(os.path.join(mtStationPath, 'WP*.*'),  recursive=True)
#binList
for i in range(len(binList)):
  splitBinList = binList[i].split('\\')
  if splitBinList[len(splitBinList) - 1].find('AVG') == -1 and splitBinList[len(splitBinList) - 1].find('dmp') == -1 and splitBinList[len(splitBinList) - 1].find('edi') == -1:
   fileListing = fileListing + '\t\t\t\t\t\t' + splitBinList[len(splitBinList) - 1] + '\n'

print ('File ListingfileListing:\n' + fileListing)

File ListingfileListing:
						BF6-9621.RSP
						BF6-9624.RSP
						BF6-9625.RSP
						EF-9515X.RSP
						EF-9515Y.RSP



In [27]:
#Now finally add the processed ASCII text files to the list
txtList = glob.glob(os.path.join(mtStationPath, '*.txt'),  recursive=True)
#txtList
for i in range(len(txtList)):
  splitTxtList = txtList[i].split('\\')
  fileListing = fileListing + '\t\t\t\t\t\t' + splitTxtList[len(splitTxtList) - 1] + '\n'

print ('File ListingfileListing:\n' + fileListing)

File ListingfileListing:
						BF6-9621.RSP
						BF6-9624.RSP
						BF6-9625.RSP
						EF-9515X.RSP
						EF-9515Y.RSP
						readme.txt
						USA-Arkansas-Buffalo_River-2017-AMT050-FC6_01.txt
						USA-Arkansas-Buffalo_River-2017-AMT050-FC6_02.txt
						USA-Arkansas-Buffalo_River-2017-AMT050-FC6_03.txt
						USA-Arkansas-Buffalo_River-2017-AMT050-FC7_01.txt
						USA-Arkansas-Buffalo_River-2017-AMT050-FC7_02.txt
						USA-Arkansas-Buffalo_River-2017-AMT050-FC7_03.txt
						USA-Arkansas-Buffalo_River-2017-AMT050-FC8_01.txt
						USA-Arkansas-Buffalo_River-2017-AMT050-FC8_02.txt
						USA-Arkansas-Buffalo_River-2017-AMT050-FC9_01.txt
						USA-Arkansas-Buffalo_River-2017-AMT050-FC9_02.txt
						USA-Arkansas-Buffalo_River-2017-AMT050-FCA_AA.txt
						USA-Arkansas-Buffalo_River-2017-AMT050-FCB_AB.txt
						USA-Arkansas-Buffalo_River-2017-AMT050-FCC_AC.txt
						USA-Arkansas-Buffalo_River-2017-AMT050-FCC_AD.txt
						USA-Arkansas-Buffalo_River-2017-AMT050-FCD_AE.txt
						USA-Arkansas-Buff

# Now get values and stats on the .RSP files that are listed to add to the Ent. and Att. information

### These files are all fixed width format (666) but it looks like the BFS*.RSP and EF*.RSP are slightly different beasts that have different formats

In [28]:
# Load the BFS*.RSP files into pandasand create entities and attributes
allRspEandP = ''
strFreqMaxBSP = ''
strAmpMaxBSP = ''
strGammaMaxBSP = ''
strFreqMinBSP = ''
strAmpMinBSP = ''
strGammaMinBSP = ''

strFreqMaxESP = ''
strFreqMinESP = ''
strAmp1MaxESP = ''
strAmp1MinESP = ''
strAmp2MaxESP = ''
strAmp2MinESP = ''
strAmp3MaxESP = ''
strAmp3MinESP = ''
strAmp4MaxESP = ''
strAmp4MinESP = ''
strPhz1MaxESP = ''
strPhz1MinESP = ''
strPhz2MaxESP = ''
strPhz2MinESP = ''
strPhz3MaxESP = ''
strPhz3MinESP = ''
strPhz4MaxESP = ''
strPhz4MinESP = ''

intNumberBSF = 0
intNumberEFF = 0
for i in range(len(rspList)):
#for i in range(3):
    if rspList[i].find('BF') > 0: #Do this if the file has a BF style Fixed Width Format
        bfRSP = pd.read_fwf(rspList[i], widths=[6,6,6], skiprows=5, parse_dates=True).rename(columns={'31':'Freq', '1':'Amp', 'Unnamed: 2':'Gamma'})
#print (rspList[1])
        print (bfRSP)    
    
# Now lets get the stats of the bfRSP data

# Make Array of Min and Max Values
        FreqMaxRSP = bfRSP['Freq'].max()
        strFreqMaxBSP = str(FreqMaxRSP)
        AmpMaxRSP = bfRSP['Amp'].max()
        strAmpMaxBSP = str(AmpMaxRSP)
        GammaMaxRSP = bfRSP['Gamma'].max()
        strGammaMaxBSP = str(GammaMaxRSP)
        FreqMinRSP = bfRSP['Freq'].min()
        strFreqMinBSP = str(FreqMinRSP)
        AmpMinRSP = bfRSP['Amp'].min()
        strAmpMinBSP = str(AmpMinRSP)
        GammaMinRSP = bfRSP['Gamma'].min()
        strGammaMinBSP = str(GammaMinRSP)
# now print RSP entity and attribute    
        bspEandA = '\t\t<detailed>\n\t\t\t<enttypl>Text File ' + rspFileListing [i] + '</enttypl>\n' \
        + '\t\t\t<enttypd>System Calibration File</enttypd>\n\t\t\t<enttypds>Electromagnetic Instruments (EMI)</enttypds>\n\t\t</enttyp>\n' \
        + '\t\t\t<attr>\n\t\t\t<attrlabl>Freq</attrlabl>\n\t\t\t<attrdef>Frequency - Hz</attrdef>\n\t\t\t<attrdefs>U.S. Geological Survey</attrdefs>' \
        + '\n\t\t\t<attrdomv>\n\t\t\t\t<rdom>Frequency - Hz</rdom>\n\t\t\t\t<rdommin>' \
        + strFreqMinBSP + '</rdommin>\n\t\t\t\t<rdommax>' + strFreqMaxBSP + '</rdommax>\n\t\t\t\t' \
        + '<attrunit>Hz</attrunit>\n\t\t\t</attrdomv>\n\t\t</attr>\n' \
        + '\t\t\t<attr>\n\t\t\t<attrlabl>Amp</attrlabl>\n\t\t\t<attrdef>Amplitude - Volts/Gamma</attrdef>\n\t\t\t<attrdefs>U.S. Geological Survey</attrdefs>' \
        + '\n\t\t\t<attrdomv>\n\t\t\t\t<rdom>Amplitude - Volts/Gamma</rdom>\n\t\t\t\t<rdommin>' \
        + strAmpMinBSP + '</rdommin>\n\t\t\t\t<rdommax>' + strAmpMaxBSP + '</rdommax>\n\t\t\t\t</rdommax>\n\t\t\t\t' \
        + '<attrunit>Volts/Gamma</attrunit>\n\t\t\t</attrdomv>\n\t\t</attr>\n' \
        + '\t\t\t<attr>\n\t\t\t<attrlabl>Phz</attrlabl>\n\t\t\t<attrdef>Phase - Degrees</attrdef>\n\t\t\t<attrdefs>U.S. Geological Survey</attrdefs>' \
        + '\n\t\t\t<attrdomv>\n\t\t\t\t<rdom>Phase - Degrees</rdom>\n\t\t\t\t<rdommin>' \
        + strGammaMinBSP + '</rdommin>\n\t\t\t\t<rdommax>' + strGammaMaxBSP + '</rdommax>\n\t\t\t\t</rdommax>\n\t\t\t\t' \
        + '<attrunit>Degrees</attrunit>\n\t\t\t</attrdomv>\n\t\t</attr>\n' \
        + '\t\t</attr>\n\t</detailed>'
        print ('i Loop = ' + str(i))
        allRspEandP = allRspEandP +  bspEandA + '\n'
    if rspList[i].find('EF') > 0:
        print ('i Loop = ' + str(i) + ' EF File Found')
        efRSP = pd.read_fwf(rspList[i], widths=[6,6,6,6,6,6,6,6,6], skiprows=5, parse_dates=True)\
        .rename(columns={'42':'Freq', '4':'Amp1', 'Low F':'Phz1', 'requen':'Amp2', 'cy':'Phz2', '10Hz':'Amp3', 'Out':'Phz3',\
        'Unnamed: 7':'Amp4', 'Unnamed: 8':'Phz4'})
#Delete headers from data to remove all strings from Pandas Data Frame

        efRSP = efRSP[efRSP.Amp2 != 'Freque']

# Make Array of Min and Max Values
        strFreqMaxESP = str(efRSP['Freq'].max())
        strFreqMinESP = str(efRSP['Freq'].min())
        strAmp1MaxESP = str(efRSP['Amp1'].max())
        strAmp1MinESP = str(efRSP['Amp1'].min())
        strAmp2MaxESP = str(efRSP['Amp2'].max())
        strAmp2MinESP = str(efRSP['Amp2'].min())
        strAmp3MaxESP = str(efRSP['Amp3'].max())
        strAmp3MinESP = str(efRSP['Amp3'].min())
        strAmp4MaxESP = str(efRSP['Amp4'].max())
        strAmp4MinESP = str(efRSP['Amp4'].min())
        strPhz1MaxESP = str(efRSP['Phz1'].max())
        strPhz1MinESP = str(efRSP['Phz1'].min())
        strPhz2MaxESP = str(efRSP['Phz2'].max())
        strPhz2MinESP = str(efRSP['Phz2'].min())
        strPhz3MaxESP = str(efRSP['Phz3'].max())
        strPhz3MinESP = str(efRSP['Phz3'].min())
        strPhz4MaxESP = str(efRSP['Phz4'].max())
        strPhz4MinESP = str(efRSP['Phz4'].min())
        espEandA = '\t\t<detailed>\n\t\t\t<enttypl>Text File ' + rspFileListing [i] + '</enttypl>\n' \
        + '\t\t\t<enttypd>System Calibration File</enttypd>\n\t\t\t<enttypds>Electromagnetic Instruments (EMI)</enttypds>\n\t\t</enttyp>\n' \
        + '\t\t\t<attr>\n\t\t\t<attrlabl>Freq</attrlabl>\n\t\t\t<attrdef>Frequency - Hz</attrdef>\n\t\t\t<attrdefs>U.S. Geological Survey</attrdefs>' \
        + '\n\t\t\t<attrdomv>\n\t\t\t\t<rdom>Frequency - Hz</rdom>\n\t\t\t\t<rdommin>' \
        + strFreqMinESP + '</rdommin>\n\t\t\t\t<rdommax>' + strFreqMaxESP + '</rdommax>\n\t\t\t\t' \
        + '<attrunit>Hz</attrunit>\n\t\t\t</attrdomv>\n\t\t</attr>\n' \
        + '\t\t\t<attr>\n\t\t\t<attrlabl>Amp</attrlabl>\n\t\t\t<attrdef>Amplitude - Volts/Gamma</attrdef>\n\t\t\t<attrdefs>U.S. Geological Survey</attrdefs>' \
        + '\n\t\t\t<attrdomv>\n\t\t\t\t<rdom>Amplitude - Volts/Gamma</rdom>\n\t\t\t\t<rdommin>' \
        + strAmp1MinESP + '</rdommin>\n\t\t\t\t<rdommax>' + strAmp1MaxESP + '</rdommax>\n\t\t\t\t</rdommax>\n\t\t\t\t' \
        + '<attrunit>Volts/Gamma</attrunit>\n\t\t\t</attrdomv>\n\t\t</attr>\n' \
        + '\t\t\t<attr>\n\t\t\t<attrlabl>Phz</attrlabl>\n\t\t\t<attrdef>Phase - Degrees</attrdef>\n\t\t\t<attrdefs>U.S. Geological Survey</attrdefs>' \
        + '\n\t\t\t<attrdomv>\n\t\t\t\t<rdom>Phase - Degrees</rdom>\n\t\t\t\t<rdommin>' \
        + strPhz1MinESP + '</rdommin>\n\t\t\t\t<rdommax>' + strPhz1MaxESP + '</rdommax>\n\t\t\t\t</rdommax>\n\t\t\t\t' \
        + '<attrunit>Degrees</attrunit>\n\t\t\t</attrdomv>\n\t\t</attr>\n' \
        + '\t\t</attr>\n\t</detailed>'
        print ('i Loop = ' + str(i))
        allRspEandP = allRspEandP +  espEandA + '\n' 
        print (str(allRspEandP)) 

        
EandA = allRspEandP # [-:2] # this removes the last characters of the string to get rid of the last line return.  


        Freq      Amp  Gamma
0       0.10  0.00191   89.3
1       0.15  0.00287   88.9
2       0.20  0.00383   88.5
3       0.30  0.00574   87.8
4       0.40  0.00764   87.1
5       0.60  0.01140   85.7
6       0.80  0.01520   84.2
7       1.00  0.01890   82.8
8       1.50  0.02810   79.2
9       2.00  0.03700   75.8
10      3.00  0.05360   69.2
11      4.00  0.06820   63.1
12      6.00  0.09130   52.7
13      8.00  0.10700   44.6
14     10.00  0.11800   38.3
15     15.00  0.13600   27.4
16     20.00  0.13790   23.1
17     30.00  0.14800   14.1
18     40.00  0.14800   10.1
19     80.00  0.15000    5.6
20    100.00  0.15000    4.5
21    200.00  0.15190    3.2
22    400.00  0.15190    0.9
23   1000.00  0.15190    4.3
24   2000.00  0.15190   -0.5
25   4000.00  0.15190   -1.7
26   8000.00  0.15500   -4.4
27  10000.00  0.15700   -6.2
28  15000.00  0.15800  -10.2
29  20000.00  0.16300  -14.0
30  25000.00  0.16941  -19.6
i Loop = 0
        Freq      Amp  Gamma
0       0.10  0.00183   89.3
1  

## Now let's work on the SD mode files

### Overview of SD files,

The prefix file naming convention for the SD mode AMT non-transmitter files are:
AANNNRN where AA is the survey area, NNN is the site number, R is the run "number" (run A, run B, run C, etc.), and N is a subset run number (first run is A1, second run is A2, third run is A3, etc.)

The suffix file naming convention for the SD mode AMT non-transmitter files are:
SD6 sample frequencies are always 79, 90, 100, 150, 210, 270, 340, 460, 580 Hertz
SD7 sample frequencies are always 340, 460, 580, 720, 885, 1170 Hertz
SD8 sample frequencies are always 1170, 1500, 1870, 2200, 2730, 3550, 4900, 6500, 9000 Hertz
SD9 sample frequencies are always 6500, 9000, 11590, 15290, 19500, 23370 Hertz

The prefix file naming convention for the SD mode AMT transmitter files are:
AANNNRN where AA is the survey area, NNN is the site number, R is the run "number" (run A, run B, run C, etc.), and N is a subset run "number" (first run is AA, second run is AB, third run is AC, etc.)

The suffix file naming convention for the SD mode AMT transmitter files are:
SDA sample frequency is always 960 Hertz
SDB sample frequency is always 1200 Hertz
SDC sample frequency is always 1870 Hertz
SDD sample frequency is always 2420 Hertz
SDE sample frequency is always 2730 Hertz
SDF sample frequency is always 3550 Hertz
SDG sample frequency is always 5210 Hertz
SDH sample frequency is always 6850 Hertz
SDI sample frequency is always 11590 Hertz
SDJ sample frequency is always 15920 Hertz
SDK sample frequency is always 23370 Hertz

The file format of the SD mode files are described in each of the text files (last block of information within each file). It is the same data format regardless of whether or not a transmitter was used to collect the data. We must have been looking at the text file for the FC (Fourier Coefficient) files in your office, so I was confused that I hadn't described the data format (it is there in the text files for all SD files).



In [29]:
#First get a listing of the SD mode files
#Now finally add the processed ASCII text files to the list
sdList = glob.glob(os.path.join(mtStationPath, '*.sd?'),  recursive=True)
#sdaList
sdFileList = ''
for i in range(len(sdList)):
  splitSdList = sdList[i].split('\\')
  fileListing = fileListing + '\t\t\t\t\t\t' + splitSdList[len(splitSdList) - 1] + '\n'
  sdFileList = sdFileList + splitSdList[len(splitSdList) - 1] + '\n'

#print ('File ListingfileListing:\n' + fileListing)
print ('SD file listing:\n' + sdFileList)


SD file listing:
AR050A1.SD6
AR050A1.SD7
AR050A1.SD8
AR050A1.SD9
AR050A2.SD6
AR050A2.SD7
AR050A2.SD8
AR050A2.SD9
AR050A3.SD6
AR050A3.SD7
AR050AA.SDA
AR050AB.SDB
AR050AC.SDC
AR050AD.SDC
AR050AE.SDD
AR050AF.SDE
AR050AG.SDE
AR050AH.SDF
AR050AI.SDG
AR050AJ.SDG
AR050AK.SDH
AR050AL.SDH
AR050AM.SDI
AR050AN.SDG
AR050AO.SDH
AR050AP.SDG
AR050AQ.SDG
AR050AR.SDG
AR050AS.SDH
AR050AT.SDI
AR050AU.SDI
AR050BA.SDI
AR050BB.SDJ
AR050BC.SDJ
AR050BD.SDJ
AR050BE.SDJ
AR050BF.SDK
AR050BG.SDK
AR050BH.SDK



In [30]:
#As before, we read in the list of files

for i in range(len(sdList)):
#for i in range(3):
    #if rspList[i].find('BF') > 0: #Do this if the file has a BF style Fixed Width Format
        dfSD = pd.read_fwf(sdList[i], widths=[15,15,15,15,15], skiprows=27, parse_dates=True).rename(columns={'79.0000  32.':'Amp1', '00000   20   1':'Amp2', 'Unnamed: 2':'Amp3', 'Unnamed: 3':'Amp4', 'Unnamed: 4':'Amp5'})
#Now strip out the header rows in the data as to not screw up calculating the range
        #dfSD = dfSD[dfSD.Amp4 != '00000   20   1']
        print (dfSD)        
    

               Amp1             Amp2          Amp3          Amp4          Amp5
0   62.72856888e-02              0.0 -1.766098e-01 -1.174255e+00  2.639179e+00
1               0.0  -1.80749909e-03 -5.395094e-03  1.198498e-02 -2.294817e-03
2   57.28296150e-06              0.0 -7.695132e-03 -1.139526e-03  3.628979e-03
3   -1.53448091e-02  31.18834930e-06 -6.787768e-05  1.039011e-04  0.000000e+00
4   -8.36373704e-03  71.21806942e-04 -1.337332e-02 -1.799513e-02 -4.491714e-05
5   -9.67998803e-05  94.87897303e-06 -1.125224e-04  2.118639e-04  0.000000e+00
6      90.0000  32.   00000   20   1           NaN           NaN           NaN
7   10.38863225e-02              0.0 -1.054268e-01 -1.020941e-01  3.894446e-01
8               0.0  -5.12833551e-04 -4.437172e-04  1.757035e-03 -3.110686e-04
9   86.12946189e-07              0.0 -8.663656e-04  3.273872e-05  7.734232e-04
10  -2.08997153e-03  58.78760589e-07 -9.429497e-06  1.924893e-05  0.000000e+00
11  -4.80074602e-04  59.45507671e-05 -2.048374e-03 -

    11590.0000 620.   00000   15   1          Amp3      Amp4          Amp5
0   79.85494857e-01              0.0 -3.474947e-01 -0.317240  5.949173e-02
1               0.0  -3.61664156e-04  1.081633e-04  0.000064 -4.170635e-05
2   16.74744326e-08              0.0 -7.512569e-03  0.004708  6.599763e-05
3   -4.47351681e-04  19.93720370e-08 -4.540971e-08  0.000010  0.000000e+00
4   45.48809800e-05  18.38767691e-04  4.343061e-05 -0.000091  2.667892e-07
5   -9.56967733e-08  12.38862072e-08 -2.155448e-06  0.000001  0.000000e+00
6   11590.0000 620.   00000   15   1           NaN       NaN           NaN
7   10.36181495e+00              0.0 -4.971611e-01 -0.378713  6.400164e-02
8               0.0  -4.50211883e-04  2.847004e-04  0.000045 -3.986964e-05
9   11.64211613e-08              0.0 -9.558939e-03  0.006091  1.736413e-04
10  -6.06707761e-04  46.39141416e-08  4.919875e-08  0.000013  0.000000e+00
11  64.77412285e-05  27.94608181e-04 -2.318260e-05 -0.000151  2.372671e-07
12  -2.12413731e-07  61.8

In [31]:
#listing of the Fourier Coefficient files (FC) files.
#These are Binary so we don't need to come up with the range of the values.
fcList = glob.glob(os.path.join(mtStationPath, '*.fc?'),  recursive=True)
#fcList
for i in range(len(fcList)):
  splitFcList = fcList[i].split('\\')
  fileListing = fileListing + '\t\t\t\t\t\t' + splitFcList[len(splitFcList) - 1] + '\n'

print ('File ListingfileListing:\n' + fileListing)

File ListingfileListing:
						BF6-9621.RSP
						BF6-9624.RSP
						BF6-9625.RSP
						EF-9515X.RSP
						EF-9515Y.RSP
						readme.txt
						USA-Arkansas-Buffalo_River-2017-AMT050-FC6_01.txt
						USA-Arkansas-Buffalo_River-2017-AMT050-FC6_02.txt
						USA-Arkansas-Buffalo_River-2017-AMT050-FC6_03.txt
						USA-Arkansas-Buffalo_River-2017-AMT050-FC7_01.txt
						USA-Arkansas-Buffalo_River-2017-AMT050-FC7_02.txt
						USA-Arkansas-Buffalo_River-2017-AMT050-FC7_03.txt
						USA-Arkansas-Buffalo_River-2017-AMT050-FC8_01.txt
						USA-Arkansas-Buffalo_River-2017-AMT050-FC8_02.txt
						USA-Arkansas-Buffalo_River-2017-AMT050-FC9_01.txt
						USA-Arkansas-Buffalo_River-2017-AMT050-FC9_02.txt
						USA-Arkansas-Buffalo_River-2017-AMT050-FCA_AA.txt
						USA-Arkansas-Buffalo_River-2017-AMT050-FCB_AB.txt
						USA-Arkansas-Buffalo_River-2017-AMT050-FCC_AC.txt
						USA-Arkansas-Buffalo_River-2017-AMT050-FCC_AD.txt
						USA-Arkansas-Buffalo_River-2017-AMT050-FCD_AE.txt
						USA-Arkansas-Buff

In [32]:
# Create the Entity and Attributes for the FC listing
for i in range(len(fcList)):
    splitFcList = fcList[i].split('\\')
    fcEandA = '\t\t<detailed>\n\t\t\t<enttypl>Binary File ' + splitFcList[len(splitFcList) - 1] + '</enttypl>\n' \
        + '\t\t\t<enttypd>Raw Time Series Binary Data File. The header information for this file contains field notes including survey identifier, personnel, date/time, weather, indication of a remote MT site, additional comments and sensor configuration for the electric and magnetic field channels.</enttypd>' \
        + '\n\t\t\t<enttypds>U.S. Geological Survey</enttypds>' \
        + '<attrlabl>Data Value</attrlabl>\n\t\t\t<attrdef>Binary Data Value</attrdef>\n\t\t\t<attrdefs>U.S. Geological Survey</attrdefs>\n\t\t\t<attrdomv>'\
        + '\n\t\t</enttyp>\n\t\t\t\t<udom>Binary Data Value</udom>\n\t\t\t\t</attrdomv>' \
        + '\t\t</attr>\n\t</detailed>\n'
    print (fcEandA)
        
        
        
        
        
        

		<detailed>
			<enttypl>Binary File AR050A1.FC6</enttypl>
			<enttypd>Raw Time Series Binary Data File. The header information for this file contains field notes including survey identifier, personnel, date/time, weather, indication of a remote MT site, additional comments and sensor configuration for the electric and magnetic field channels.</enttypd>
			<enttypds>U.S. Geological Survey</enttypds><attrlabl>Data Value</attrlabl>
			<attrdef>Binary Data Value</attrdef>
			<attrdefs>U.S. Geological Survey</attrdefs>
			<attrdomv>
		</enttyp>
				<udom>Binary Data Value</udom>
				</attrdomv>		</attr>
	</detailed>

		<detailed>
			<enttypl>Binary File AR050A1.FC7</enttypl>
			<enttypd>Raw Time Series Binary Data File. The header information for this file contains field notes including survey identifier, personnel, date/time, weather, indication of a remote MT site, additional comments and sensor configuration for the electric and magnetic field channels.</enttypd>
			<enttypds>U.S. Geolo

In [33]:
#listing of the Time Series TS? files
#Now finally add the processed ASCII text files to the list
tsList = glob.glob(os.path.join(mtStationPath, '*.ts?'),  recursive=True)
#tsList
for i in range(len(tsList)):
  splitTsList = tsList[i].split('\\')
  fileListing = fileListing + '\t\t\t\t\t\t' + splitTsList[len(splitTsList) - 1] + '\n'

print ('File ListingfileListing:\n' + fileListing)

File ListingfileListing:
						BF6-9621.RSP
						BF6-9624.RSP
						BF6-9625.RSP
						EF-9515X.RSP
						EF-9515Y.RSP
						readme.txt
						USA-Arkansas-Buffalo_River-2017-AMT050-FC6_01.txt
						USA-Arkansas-Buffalo_River-2017-AMT050-FC6_02.txt
						USA-Arkansas-Buffalo_River-2017-AMT050-FC6_03.txt
						USA-Arkansas-Buffalo_River-2017-AMT050-FC7_01.txt
						USA-Arkansas-Buffalo_River-2017-AMT050-FC7_02.txt
						USA-Arkansas-Buffalo_River-2017-AMT050-FC7_03.txt
						USA-Arkansas-Buffalo_River-2017-AMT050-FC8_01.txt
						USA-Arkansas-Buffalo_River-2017-AMT050-FC8_02.txt
						USA-Arkansas-Buffalo_River-2017-AMT050-FC9_01.txt
						USA-Arkansas-Buffalo_River-2017-AMT050-FC9_02.txt
						USA-Arkansas-Buffalo_River-2017-AMT050-FCA_AA.txt
						USA-Arkansas-Buffalo_River-2017-AMT050-FCB_AB.txt
						USA-Arkansas-Buffalo_River-2017-AMT050-FCC_AC.txt
						USA-Arkansas-Buffalo_River-2017-AMT050-FCC_AD.txt
						USA-Arkansas-Buffalo_River-2017-AMT050-FCD_AE.txt
						USA-Arkansas-Buff

In [34]:
# Create the Entity and Attributes for the Time series .TS? listing
for i in range(len(tsList)):
    splitTsList = tsList[i].split('\\')
    tsEandA = '\t\t<detailed>\n\t\t\t<enttypl>Binary File ' + splitTsList[len(splitTsList) - 1] + '</enttypl>\n' \
        + '\t\t\t<enttypd>Raw Time Series Binary Data File. The header information for this file contains field notes including survey identifier, personnel, date/time, weather, indication of a remote MT site, additional comments and sensor configuration for the electric and magnetic field channels.</enttypd>' \
        + '\n\t\t\t<enttypds>U.S. Geological Survey</enttypds>' \
        + '<attrlabl>Data Value</attrlabl>\n\t\t\t<attrdef>Binary Data Value</attrdef>\n\t\t\t<attrdefs>U.S. Geological Survey</attrdefs>\n\t\t\t<attrdomv>'\
        + '\n\t\t</enttyp>\n\t\t\t\t<udom>Binary Data Value</udom>\n\t\t\t\t</attrdomv>' \
        + '\t\t</attr>\n\t</detailed>\n'
    print (tsEandA)

		<detailed>
			<enttypl>Binary File AR050A1.TS1</enttypl>
			<enttypd>Raw Time Series Binary Data File. The header information for this file contains field notes including survey identifier, personnel, date/time, weather, indication of a remote MT site, additional comments and sensor configuration for the electric and magnetic field channels.</enttypd>
			<enttypds>U.S. Geological Survey</enttypds><attrlabl>Data Value</attrlabl>
			<attrdef>Binary Data Value</attrdef>
			<attrdefs>U.S. Geological Survey</attrdefs>
			<attrdomv>
		</enttyp>
				<udom>Binary Data Value</udom>
				</attrdomv>		</attr>
	</detailed>

		<detailed>
			<enttypl>Binary File AR050A2.TS1</enttypl>
			<enttypd>Raw Time Series Binary Data File. The header information for this file contains field notes including survey identifier, personnel, date/time, weather, indication of a remote MT site, additional comments and sensor configuration for the electric and magnetic field channels.</enttypd>
			<enttypds>U.S. Geolo

In [35]:
#listing of the BP? files
#Now finally add the processed ASCII text files to the list
bpList = glob.glob(os.path.join(mtStationPath, '*.bp?'),  recursive=True)
#bpList
for i in range(len(bpList)):
  splitBpList = bpList[i].split('\\')
  fileListing = fileListing + '\t\t\t\t\t\t' + splitBpList[len(splitBpList) - 1] + '\n'

print ('File ListingfileListing:\n' + fileListing)

File ListingfileListing:
						BF6-9621.RSP
						BF6-9624.RSP
						BF6-9625.RSP
						EF-9515X.RSP
						EF-9515Y.RSP
						readme.txt
						USA-Arkansas-Buffalo_River-2017-AMT050-FC6_01.txt
						USA-Arkansas-Buffalo_River-2017-AMT050-FC6_02.txt
						USA-Arkansas-Buffalo_River-2017-AMT050-FC6_03.txt
						USA-Arkansas-Buffalo_River-2017-AMT050-FC7_01.txt
						USA-Arkansas-Buffalo_River-2017-AMT050-FC7_02.txt
						USA-Arkansas-Buffalo_River-2017-AMT050-FC7_03.txt
						USA-Arkansas-Buffalo_River-2017-AMT050-FC8_01.txt
						USA-Arkansas-Buffalo_River-2017-AMT050-FC8_02.txt
						USA-Arkansas-Buffalo_River-2017-AMT050-FC9_01.txt
						USA-Arkansas-Buffalo_River-2017-AMT050-FC9_02.txt
						USA-Arkansas-Buffalo_River-2017-AMT050-FCA_AA.txt
						USA-Arkansas-Buffalo_River-2017-AMT050-FCB_AB.txt
						USA-Arkansas-Buffalo_River-2017-AMT050-FCC_AC.txt
						USA-Arkansas-Buffalo_River-2017-AMT050-FCC_AD.txt
						USA-Arkansas-Buffalo_River-2017-AMT050-FCD_AE.txt
						USA-Arkansas-Buff

In [36]:
# Create the Entity and Attributes for the .BP? listing

#As before, we read in the list of files

for i in range(len(bpList)):
#for i in range(3):
    #if rspList[i].find('BF') > 0: #Do this if the file has a BF style Fixed Width Format
        dfBP = pd.read_fwf(bpList[i], widths=[15,15,15,15,15], skiprows=27, parse_dates=True).rename(columns={'4.3945   1.':'Amp1', '85697    1   2':'Amp2', 'Unnamed: 2':'Amp3', 'Unnamed: 3':'Amp4', 'Unnamed: 4':'Amp5'})
#Now strip out the header rows in the data as to not screw up calculating the range
        #dfSD = dfSD[dfSD.Amp4 != '00000   20   1']
        print (dfBP)        


for i in range(len(bpList)):
    splitBpList = tsList[i].split('\\')
    bpEandA = '\t\t<detailed>\n\t\t\t<enttypl>Binary File ' + splitBpList[len(splitBpList) - 1] + '</enttypl>\n' \
        + '\t\t\t<enttypd>Raw Time Series Binary Data File. The header information for this file contains field notes including survey identifier, personnel, date/time, weather, indication of a remote MT site, additional comments and sensor configuration for the electric and magnetic field channels.</enttypd>' \
        + '\n\t\t\t<enttypds>U.S. Geological Survey</enttypds>' \
        + '<attrlabl>Data Value</attrlabl>\n\t\t\t<attrdef>Binary Data Value</attrdef>\n\t\t\t<attrdefs>U.S. Geological Survey</attrdefs>\n\t\t\t<attrdomv>'\
        + '\n\t\t</enttyp>\n\t\t\t\t<udom>Binary Data Value</udom>\n\t\t\t\t</attrdomv>' \
        + '\t\t</attr>\n\t</detailed>\n'
    print (bpEandA)

               Amp1             Amp2          Amp3          Amp4          Amp5
0   10.15611009e-03              0.0 -1.564040e-02 -4.240838e-03  3.809019e-02
1               0.0  -1.32593208e-04 -2.062578e-05  3.967518e-04 -6.254733e-05
2   10.04175498e-06              0.0  9.014170e-05  6.211137e-05 -2.996556e-04
3   27.45190807e-06  -4.97454458e-06  1.611817e-06  6.634777e-06  0.000000e+00
4   -4.30471593e-05  -8.39445000e-05  8.199664e-05  1.266369e-04 -3.654625e-07
5   49.41630696e-10  -1.25065746e-06 -1.175490e-07  8.588850e-06  0.000000e+00
6       7.3242   3.   83474    1   4           NaN           NaN           NaN
7   29.95576506e-03              0.0 -3.368590e-02 -6.135633e-03  7.328164e-02
8               0.0  -2.87033280e-04 -2.067439e-05  6.251969e-04 -1.022123e-04
9   63.31697223e-07              0.0  1.671871e-05  1.566712e-04 -2.644972e-04
10  47.72681482e-06  -2.79622190e-06  3.112059e-07  4.967974e-06  0.000000e+00
11  54.75258891e-06  -1.62263900e-05  7.280404e-05  

# Populate Metadata Template

In [37]:
#Load XML Metadata Template File and Read It
metaData = os.path.join(mtMataDataTemplatePath, mtMataDataTemplateName)
xmlTemplateFile = open(metaData, 'r')
metaDataContent = xmlTemplateFile.readlines()
print(metaDataContent)
xmlTemplateFile.close()


['<?xml version="1.0" encoding="UTF-8"?>\n', '<!--Record created using beta version 1.0 of the USGS GGGSC AMT Metadata Creation Tool. (https://github.com/pbrown-usgs/SquamataAssemblyAMT)-->\n', '<metadata>\n', '\t<idinfo>\n', '\t\t<citation>\n', '\t\t\t<citeinfo>\n', '\t\t\t\t<origin>Rodriguez, B. D.</origin>\n', '\t\t\t\t<origin>Hudson, M. R.</origin>\n', '\t\t\t\t<origin>Brown, P. J.</origin>\n', '\t\t\t\t<pubdate>2019</pubdate>\n', '\t\t\t\t<title>{title}</title>\n', '\t\t\t\t<edition>1</edition>\n', '\t\t\t\t<geoform>ASCII and Binary Digital Data</geoform>\n', '\t\t\t\t<pubinfo>\n', '\t\t\t\t\t<pubplace>Denver, CO</pubplace>\n', '\t\t\t\t\t<publish>U.S. Geological Survey</publish>\n', '\t\t\t\t</pubinfo>\n', '\t\t\t\t<othercit>Additional information about Originators:Rodriguez, B.D., http://orcid.org/0000-0002-2263-611X; Brown, P.J., http://orcid.org/0000-0002-2415-7462</othercit>\n', '\t\t\t\t<onlink>{onlink}</onlink>\n', '\t\t\t</citeinfo>\n', '\t\t</citation>\n', '\t\t<descript>

In [38]:
# Replace values of current metadata template with the appropriate values.  
# All of this input should have been defined when going through the steps outlined above.
lineString = ''
newMetaDataContent = metaDataContent
splitFileName = ediList[0].split('.')
myfilename = splitFileName[0] + '.xml'
xmlFile = open(myfilename,"w+")
print(myfilename)
#print(keywords.value)
for i in range(len(metaDataContent)):
    lineString = metaDataContent[i]
    if lineString.find('{title}'):
     lineString = lineString.replace('{title}', citTit + '; Station ' + drTitle)
    
    if lineString.find('{abstract}'):
     lineString = lineString.replace('{abstract}', purposeClean)
    
    if lineString.find('{purpose}'):
     lineString = lineString.replace('{purpose}', descriptionClean)
    
    if lineString.find('{BeginFileListingHere}'):
     lineString = lineString.replace('{BeginFileListingHere}', fileListing)
    
    if lineString.find('{keywords}'):
     lineString = lineString.replace('{keywords}', keywords.value)
    
    if lineString.find('{begdate}'):
     lineString = lineString.replace('{begdate}', begdate)
    
    if lineString.find('{enddate}'):
     lineString = lineString.replace('{enddate}', enddate)
    
    if lineString.find('{SiteLon}'):
     lineString = lineString.replace('{SiteLon}', sitLon)
    
    if lineString.find('{SiteLon}'):
     lineString = lineString.replace('{SiteLat}', sitLat)
    
    if lineString.find('{EandA}'):
     lineString = lineString.replace('{EandA}', EandA)
   
    # {county}

    else:
     lineString = lineString
    xmlFile.write(lineString)
    
    #print (lineString)
     
    
    
#for r in (metaDataContent):
    #newMetaDataContent = metaDataContent.replace('{title}', drTitle)
    #newMetaDataContent = metaDataContent.replace('{keywords}', keywords.value)
xmlFile.close()

print ('Creation of new metadata file is complete\n\n') 
#Load EDI File and Read It
##checkFile = open(open(myfilename, 'r')
##checkFileContent = checkFile.read()
##checkFile.close()
##print(checkFileContent)

C:\CurrentWork\DataReleases\Arkansas AMT data release\AMT050\USA-Arkansas-Buffalo_River-2017-AMT050.xml
Creation of new metadata file is complete




### At this point the new metadata file should be created.  Check the result below.  If the created file is OK, run the loop for the rest of the children...

In [39]:
# Show the resulting child xml metadata file example 
#for i in range(len(newMetaDataContent)):
print (newMetaDataContent)

['<?xml version="1.0" encoding="UTF-8"?>\n', '<!--Record created using beta version 1.0 of the USGS GGGSC AMT Metadata Creation Tool. (https://github.com/pbrown-usgs/SquamataAssemblyAMT)-->\n', '<metadata>\n', '\t<idinfo>\n', '\t\t<citation>\n', '\t\t\t<citeinfo>\n', '\t\t\t\t<origin>Rodriguez, B. D.</origin>\n', '\t\t\t\t<origin>Hudson, M. R.</origin>\n', '\t\t\t\t<origin>Brown, P. J.</origin>\n', '\t\t\t\t<pubdate>2019</pubdate>\n', '\t\t\t\t<title>{title}</title>\n', '\t\t\t\t<edition>1</edition>\n', '\t\t\t\t<geoform>ASCII and Binary Digital Data</geoform>\n', '\t\t\t\t<pubinfo>\n', '\t\t\t\t\t<pubplace>Denver, CO</pubplace>\n', '\t\t\t\t\t<publish>U.S. Geological Survey</publish>\n', '\t\t\t\t</pubinfo>\n', '\t\t\t\t<othercit>Additional information about Originators:Rodriguez, B.D., http://orcid.org/0000-0002-2263-611X; Brown, P.J., http://orcid.org/0000-0002-2415-7462</othercit>\n', '\t\t\t\t<onlink>{onlink}</onlink>\n', '\t\t\t</citeinfo>\n', '\t\t</citation>\n', '\t\t<descript>